### PR TITLE
Tutorial for Kafka Input Operator

### PR DESCRIPTION
**Pull request by chaithu14**

### CODE REVIEW

File: docs/operators/kafkaInputOperator.md [hunk 1]
```diff
@@ -0,0 +1,344 @@
+KAFKA INPUT OPERATOR
+=====================
+
+### Introduction: About Kafka Input Operator
+
+Kafka Input Operator is an fault-tolerance and scalable Malhar Operator.
+This is an input operator that consume data from Kafka messaging system
```
Comment (gauravgopi123) [inline docs/operators/kafkaInputOperator.md]: This should be the first line.. fault tolerance and scalability are it's features

File: docs/operators/kafkaInputOperator.md [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+
+Kafka is a distributed publish subscribe messaging system. Since kafka
+is distributed system, topics are partitioned and replicated across
+nodes. Kafka input operator is needed when you want to read data from N
+partitions of kafka topic in parallel in Apex application.
+
+### AbstractKafkaInputOperator
+
+This is the abstract implementation that serves as consuming messages
+from Kafka messaging system. This class doesn’t have any ports.
```
Comment (gauravgopi123) [inline docs/operators/kafkaInputOperator.md]: This is an abstract implementation that serves as base class for consuming messages from Kafka

File: docs/operators/kafkaInputOperator.md [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+
+### Introduction: About Kafka Input Operator
+
+Kafka Input Operator is an fault-tolerance and scalable Malhar Operator.
+This is an input operator that consume data from Kafka messaging system
+for further processing in Apex.
+
+### Why is it needed ?
+
+Kafka is a distributed publish subscribe messaging system. Since kafka
```
Comment (gauravgopi123) [inline docs/operators/kafkaInputOperator.md]: Since kafka is distributed system.. It is duplicate. Merge it with first line

Comment (siyuanh) [inline docs/operators/kafkaInputOperator.md]: add pull-based

File: docs/operators/kafkaInputOperator.md [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+<td align="left"><p>Parameter</p></td>
+<td align="left"><p>Type</p></td>
+<td align="left"><p>Default</p></td>
+<td align="left"><p>Description</p></td>
+</tr>
+<tr class="even">
+<td align="left"><p>maxTuplesPerWindow</p></td>
+<td align="left"><p>int</p></td>
+<td align="left"><p>MAX_VALUE</p></td>
+<td align="left"><p>Controls the number of messages emitting in each application window from this operator. Minimum value is 1.</p></td>
```
Comment (siyuanh) [inline docs/operators/kafkaInputOperator.md]: should be streaming window

File: docs/operators/kafkaInputOperator.md [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+This is an abstract implementation of kafka consumer. It sends the fetch
+requests to the leading brokers of kafka partitions. For each request,
+it receives the set of messages and store these into the buffer which is
+array blocking queue. SimpleKafkaConsumer which extends from
+KafkaConsumer and serves the functionality of Simple Kafka API and
+HighLevelKafkaConsumer which extends from KafkaConsumer and  serves the
+functionality of High level API.
+
+This operator referred the Kafka Consumer API and the version is
+0.8.1.1. It should work with any 0.8.x version of Apache Kafka.
```
Comment (siyuanh) [inline docs/operators/kafkaInputOperator.md]: And 0.7 as well, kafka support 1 major version backward compatibility

File: docs/operators/kafkaInputOperator.md [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+<td align="left"><p>Parameter</p></td>
+<td align="left"><p>Type</p></td>
+<td align="left"><p>Default</p></td>
+<td align="left"><p>Description</p></td>
+</tr>
+<tr class="even">
+<td align="left"><p>zookeeper</p></td>
+<td align="left"><p>String</p></td>
+<td align="left"><p></p></td>
+<td align="left"><p>Specifies the zookeeper quorum of Kafka clusters that you want to consume messages from. zookeeper  is a string in the form of hostname1:port1,hostname2:port2,hostname3:port3  where hostname1,hostname2,hostname3 are hosts and port1,port2,port3 are ports of zookeeper server.  If the topic name is same across the kafka clusters and want to consume data from these clusters, then configure the zookeeper as follows: c1::hs1:p1,hs2:p2,hs3:p3;c2::hs4:p4,hs5:p5,c3::hs6:p6</p>
```
Comment (siyuanh) [inline docs/operators/kafkaInputOperator.md]: Cluster name is optional

**Comment by chaithu14** [comment]
Updated the doc with Ram, Siyuan, Gaurav review comments.

File: docs/operators/kafkaInputOperator.md [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+<td align="left"><p>Parameter</p></td>
+<td align="left"><p>Type</p></td>
+<td align="left"><p>Default Value</p></td>
+<td align="left"><p>Description</p></td>
+</tr>
+<tr class="even">
+<td align="left"><p>maxTuplesPerWindow</p></td>
+<td align="left"><p>int</p></td>
+<td align="left"><p>MAX_VALUE</p></td>
+<td align="left"><p>Controls the number of messages emitted in each streaming window from this operator. Minimum value is 1.</p></td>
```
Comment (tushargosavi) [inline docs/operators/kafkaInputOperator.md]: Controls the maximum number of messages

File: docs/operators/kafkaInputOperator.md [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+<td align="left"><p>ONE_TO_ONE</p></td>
+<td align="left"><p>Operator supports two types of partitioning strategies, ONE_TO_ONE and ONE_TO_MANY.</p>
+<p>ONE_TO_ONE: If this is enabled, the AppMaster creates one input operator instance per Kafka topic partition. So the number of Kafka topic partitions equals the number of operator instances.</p>
+<p>ONE_TO_MANY: The AppMaster creates K = min(initialPartitionCount, N) Kafka input operator instances where N is the number of Kafka topic partitions. If K is less than N, the remaining topic partitions are assigned to the K operator instances in round-robin fashion. If K is less than initialPartitionCount, the AppMaster creates one input operator instance per Kafka topic partition. For example, if initialPartitionCount = 5 and number of Kafka partitions(N) = 2 then AppMaster creates 2 Kafka input operator instances.</p></td>
+</tr>
+<tr class="odd">
+<td align="left"><p>msgRateUpperBound</p></td>
+<td align="left"><p></p></td>
+<td align="left"><p></p></td>
+<td align="left"><p>Maximum messages upper bound. Operator repartitions when the *msgProcessedPS* exceeds this bound.</p></td>
```
Comment (tushargosavi) [inline docs/operators/kafkaInputOperator.md]: what is msgProcessedPS?

File: docs/operators/kafkaInputOperator.md [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+void emitTuple(Message message): Abstract method that emits tuples
+extracted from Kafka message.
+
+### KafkaConsumer
+
+This is an abstract implementation of Kafka consumer. It sends the fetch
+requests to the leading brokers of Kafka partitions. For each request,
+it receives the set of messages and store them into the buffer which is
+ArrayBlockingQueue. SimpleKafkaConsumer which extends
+KafkaConsumer and serves the functionality of Simple Kafka API and
```
Comment (tushargosavi) [inline docs/operators/kafkaInputOperator.md]: Simple Consumer API

**Comment by chaithu14** [comment]
Incorporated the review comments.

**Comment by chaithu14** [comment]
@amberarrow : Incorporated the review comments. Please merge this PR.

**Pull request merged**